File: solutions/Linked List/206_reverse-linked-list.py
from typing import Optional
from linkedlist import ListNode

class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:

        prev, curr = None, head

        while curr:
            temp = curr.next
            curr.next = prev
            prev = curr
            curr = temp

        return prev
    
# Time Complexity: O(n)
# Space Complexity: O(1)


class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head or not head.next:
            return head

        prev = head
        curr = head.next
        prev.next = None

        while curr:
            temp = curr.next
            curr.next = prev
            prev = curr
            curr = temp

        return prev
    
# Time Complexity: O(n)
# Space Complexity: O(1)

File: C++/Linked List/206_reverse-linked-list.cpp
#include "linkedlist.hpp"
#include <stack>

class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        stack<int> stk;
        ListNode* curr = head;
        while (curr != nullptr) {
            stk.push(curr->val);
            curr = curr->next;
        }

        curr = head;
        while (curr != nullptr) {
            curr->val = stk.top(); stk.pop();
            curr = curr->next;
        }

        return head;
    }
};


class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        ListNode* curr = head;
        ListNode* prev = nullptr;
        ListNode* temp;
        while (curr != nullptr) {
            temp = curr->next;
            curr->next = prev;
            prev = curr;
            curr = temp;
        }

        return prev;
    }
};